query/query.go

```diff
@@ -0,0 +1,169 @@
+package query
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/fabric8-services/fabric8-wit/errors"
+	"github.com/fabric8-services/fabric8-wit/gormsupport"
+	"github.com/fabric8-services/fabric8-wit/log"
+	"github.com/goadesign/goa"
+	"github.com/jinzhu/gorm"
+	uuid "github.com/satori/go.uuid"
+)
+
+// APIStringTypeQuery helps to avoid string literal
+const APIStringTypeQuery = "queries"
+
+// Query describes a single Label
+type Query struct {
+	gormsupport.Lifecycle
+	ID      uuid.UUID `sql:"type:uuid default uuid_generate_v4()" gorm:"primary_key"` // This is the ID PK field
+	SpaceID uuid.UUID `sql:"type:uuid"`
+	Creator uuid.UUID `sql:"type:uuid"`
+	Title   string
+	Fields  string
+}
+
+// GetETagData returns the field values to use to generate the ETag
+// func (q Query) GetETagData() []interface{} {
+// 	return []interface{}{q.ID, q.Version}
+// }
+
+// GetLastModified returns the last modification time
+func (q Query) GetLastModified() time.Time {
+	return q.UpdatedAt.Truncate(time.Second)
+}
+
+// TableName overrides the table name settings in Gorm to force a specific table name
+// in the database.
+func (q Query) TableName() string {
+	return "queries"
+}
+
+// Repository describes interactions with Labels
+type Repository interface {
+	Create(ctx context.Context, u *Query) error
+	List(ctx context.Context, spaceID uuid.UUID) ([]Query, error)
+	ListByCreator(ctx context.Context, spaceID uuid.UUID, creatorID uuid.UUID) ([]Query, error)
+	Load(ctx context.Context, queryID uuid.UUID, spaceID uuid.UUID) (*Query, error)
+	Delete(ctx context.Context, ID uuid.UUID) error
+}
+
+// NewQueryRepository creates a new storage type.
+func NewQueryRepository(db *gorm.DB) Repository {
+	return &GormQueryRepository{db: db}
+}
+
+// GormQueryRepository is the implementation of the storage interface for Labels.
+type GormQueryRepository struct {
+	db *gorm.DB
+}
+
+// GetETagData returns the field values to use to generate the ETag
+func (q Query) GetETagData() []interface{} {
+	return []interface{}{q.ID, strconv.FormatInt(q.UpdatedAt.Unix(), 10)}
```

Comment:
Is there any particular reason to break with existing common practice of having a `Version` field and using that i for calculating the etag?

Reply:
we are not supporting `Update` on queries yet, so I have not introduced `Version` field on the table.
Hence I used updated_at time.
I see similar implementation for few other packages one of them is `Identity`

Reply:
Okay.